### ISSUE TITLE
[Question] What is interior mutability and why Rust is good even if it finds an error on runtime?

### ISSUE THREAD

**Issue by integerblue** [kaist-cp/cs492-concur#61]
At first, I am confused about interior mutability.
I don't understand the statement that users are using the safe API as if there are no SMAs. 
RefCell was the example for interior mutability, but it said the error comes out in runtime.
I thought a 'safe' system should not allow error a runtime error.
You would not be able to figure out the error if the runtime error comes out stochastically, and the bug could come out after deployment.
Why is RefCell a safe API even if it uses unsafe code and why is that an example of interior mutability?

**Comment by tomtomjhj** [comment]
* Informal definition of safety in Rust: https://doc.rust-lang.org/nomicon/what-unsafe-does.html
* On safe API (from [RustBelt paper] §1.1)
  > [..] the implementations of Rust’s standard libraries make widespread use of unsafe operations, such as “raw pointer” manipulations for which aliasing is not tracked. The developers of these libraries claim that their uses of unsafe code have been properly “encapsulated”, meaning that if programmers make use of the APIs exported by these libraries but otherwise avoid the use of unsafe operations themselves, then their programs should never exhibit any unsafe/undefined behaviors.
* On interior mutability (from [RustBelt paper] §2.5)
  > [..] Rust ensures memory safety and data-race freedom by ruling out the combination of aliasing and mutation. However, there are cases where shared mutable state is actually needed to (efficiently) implement an algorithm or a data structure. To support these usecases, Rust provides some primitives providing shared mutable state. All of these have in common that they permit mutation through a shared reference—a concept called interior mutability.
  >
  > At this point, you may be wondering—how does this fit together with the story of mutation and aliasing being the root of all memory and thread safety problems? The key point is that these primitives have a carefully controlled API surface. Even though mutation through a shared reference is unsafe in general, it can still be safe when appropriate restrictions are enforced by either static or run-time checks.

[RustBelt paper]: https://plv.mpi-sws.org/rustbelt/popl18/

**Comment by YujinGaya** [comment]
It's safe in a way that it retains the correctness of the program and if it failed to do so, it halts. If we allowed concurrent mutation to the shared resources, it would easily end up being an undefined state.

Assume you have a counter variable with value 0 and you add 1 to it from two threads. If data race happens, you would silently get 1 instead of the desired 2. If we used RefCell api, instead of getting 1, our program would panic and we can fix it in advance to the deployment, or if we already deployed, we would know that something bad happened. Imagine the counter was counting the number of precious items our customer ordered. If you paid for two (somehow concurrently, maybe one with a phone and one with a laptop at the same time?) and got only one, it will be a frustrating experience.

A more non-trivial example would be enums, or union in C. Imagine we have two variants with different fields. (of course, this is not how Rust stores socket address):

```Rust
enum SocketAddr {
    V4 { address: [u8; 4], port: u16},
    V6 { address: [u8; 16], port: u16},
}
```

with memory layout like following:

<table>
  <thead>
  <tr>
    <td></td>
    <td>0</td>
    <td>8</td>
    <td>16</td>
    <td>24</td>
    <td>32</td>
    <td>40</td>
    <td>...</td>
  </tr> 
  </thead>
  <tbody>
  <tr>
    <td>V4</td>
    <td colspan="4"> address </td>
    <td colspan="2"> port </td>
    <td> ... </td>
  </tr>
  <tr>
    <td>V6</td>
    <td colspan="16"> address </td>
  </tbody>
</table>

Assume we have two threads sharing a pointer to `SocketAddr` instance initialized with `V4 { address: [0; 4], port: 8080 }`. If a race condition occurs and one thread overwrites the instance with `V6 { address: [0; 16], port: 6636 }`, and other thread not knowing what's happening write `65535` to the `port` field thinking it is still V4, it will end up messy state. It looks like a V6 variant but part of its address is from V4 port number that other thread wrote. It will be `V6 { address: [0, 0, 0, 0, 255, 255, 0, ... , 0], port: 6636 }` that points nowhere. Problem is that the program doesn't panic, so this subtle error will propagate until it causes failure and we won't know how the failure happens. If we used RefCell api, the program would instantly panic and we would know that we did something wrong and specifically in what file, in which line.

> I thought a 'safe' system should not allow error a runtime error.

I think a 'safe' system should halt whenever there's no way to keep the correct behavior!
